rfcs/2021-07-19-8216-multiple-pipelines.md

```diff
@@ -0,0 +1,253 @@
+
+If any of the following constraints are violated, Vector will error on boot:
+
+- There cannot be several pipelines with the same id (for example `load-balancer.yml` and `load-balancer.json`).
+- The pipeline's configuration files should only contain transforms.
+- A pipeline's transform cannot have the same name as any component from the root configuration.
+- A pipeline cannot use another pipeline's component as input or output.
+
+If the violation occurs during a reload, an error will be triggered and handled in the same fashion as other reload errors.
+
+### Implementation
+
+#### Internal representation
+
+As mentioned in the previous section, a pipeline is _just_ a set of transforms.
+
+To be able to forward the events going through the pipeline to a `sink`, we'll add a new option `outputs` on the pipeline's transforms that will simply specify where the transforms events are redirected to.
+
+`outputs` options are used solely to build the topology and represent an interface between the transform and the external sinks.
+
+A pipeline will have the following internal representation before building the topology.
+
+```rust
+struct PipelineTransform {
+  inner: TransformOuter,
+  outputs: Vec<String>,
+}
+
+struct Pipeline {
+  id: String,
+  transforms: Map<String, PipelineTransform>,
+}
+```
+
+Which corresponds to the following configuration file:
+
+```toml
+# /etc/vector/pipelines/pipeline.toml
+[transforms.foo]
+type = "remap"
+inputs = ["from-root"]
+outputs = ["dc1", "dc2"]
+# ...
+
+[transforms.bar]
+type = "remap"
+inputs = ["foo"]
+outputs = ["dc-us", "dc-eu"]
+# ...
+```
+
+The `outputs` option is made to forward the events from inside the pipeline to an external component.
+
+#### From configuration to topology
+
+If we look deeper at the configuration building process, the configuration compiler will require the pipelines to build the [configuration](https://github.com/timberio/vector/blob/v0.15.0/src/config/builder.rs#L71).
+
+To do so, we'll need to implement a `Pipeline` from the previous section. We'll then update [the `compile` function](https://github.com/timberio/vector/blob/v0.15.0/src/config/compiler.rs#L4) to build a `Config` containing the required pipelines components. The compiler will load the pipelines' transforms and add the `outputs` to the corresponding `sinks`.
+
+The components coming from the pipeline will be cloned inside the final `Config`, in the `IndexMap` containing the `transforms` and the `outputs` from the pipeline components will be added to the referring components input field.
+
+For example, the following configuration and pipeline, and its **equivalent** once built.
+
+```toml
+# /etc/vector/vector.toml
+[sources.in]
+# ...
+
+[sinks.out]
+# ...
+
+# /etc/vector/pipelines/foo.toml
+[transforms.bar]
+inputs = ["in"]
+outputs = ["out"]
+# ...
+
+# equivalent once compiled
+[sources.in]
+# ...
+
+[transforms.foo#baz]
```

Comment:
Should this `bar` instead of `baz`?

Reply:
The `#` is mentioned below:

> In order to avoid internal conflicts with the pipeline components `id`s, the components `id`s will be internally prefixed with the pipeline `id` (`pipeline_id#component_id`) but the user can still reference the components with their `id` inside the pipeline configuration.

I am wondering about this too though.

Reply:
Are we envisioning that the pipeline components would be re-ID'd to `foo#baz` (component `baz` in pipeline `foo`)? These IDs are used in the tracing spans, log messages, etc..

We'll also likely want to document that users shouldn't use the `#` in their component names.

At least for metrics, I think we should avoid the prefix given that `pipeline_id` is a separate tag.

Reply:
My initial intention was to internally prefix it with the `pipeline_id` only during the building of the `Topology`. Before being built, the pipeline's transforms still have there own id, once the topology, they still have there own id. This was just an attempt to treat global transforms and pipeline's transforms the same way. But I realize this can be done without event doing that.
We'd have to update the [`Pieces` structure](https://github.com/timberio/vector/blob/master/src/topology/builder.rs#L25) in order to have a new attribute `pipelines: HashMap<String, HashMap<String, Task>>`.
Also, the [`ConfigDiff` structure](https://github.com/timberio/vector/blob/master/src/config/diff.rs#L5) will have to become a bit more complicated and do the inner of each pipeline.
In the end, it's doable without this trick, although it might take a bit more time to refactor/mutualize the existing code to work with Pipelines.
I'll update the RFC accordingly 😉

Reply:
One alternative could be to introduce something like `struct TaskId { pipeline: Option<String>, id: String }` and use that as the key for the maps in `Pieces` instead of just a raw string.

Reply:
I am starting to think maybe we do want to model it explicitly in the `TaskId`. This'll require some refactoring, as you note, but should make interacting with component ids more explicit. This could then also be used to easily expose this information in the GraphQL API without needing to do string splitting.

I say we give that a shot first at least.

Reply:
@leebenson had some arguments about that I think.

Reply:
My main thought here was that we need some mechanism to determine that a component is _internal_ , vs. one that has been explicitly configured by a user in top-level config.

Along with pipelines, this would also be useful for the Datadog observability feature I'm working on now, where an API key effectively creates an 'internal' pipeline outside of top-level config.

This flag would be useful for avoiding name clashes and having the GraphQL API make clear the distinction between what is configured by the user and created via some internal magic.

Reply:
After discussing with @leebenson, we ended up with the following proposal.
In the `ConfigBuilder` and `Topology` builder, instead of having `HashMap<String, Component>`, we should update with a `Hash<ComponentName, Component>` with:

```rust
enum ComponentName {
  Public { name: String },
  Scoped { scope: String, name: String },
  Internal { name: String },
}
```